Sources/DangerDependenciesResolver/PackageDataProvider.swift

```diff
@@ -0,0 +1,117 @@
+import DangerShellExecutor
+import Foundation
+import Logger
+import Version
+
+protocol PackageDataProviding {
+    func nameOfPackage(at url: URL, temporaryFolder: String) throws -> String
+    func latestMajorVersionForPackage(at url: URL) throws -> Int
+    func resolvePinnedPackages(generatedFolder: String) throws -> [Package.Pinned]
+}
+
+struct PackageDataProvider: PackageDataProviding {
+    let fileReader: FileReading
+    let logger: Logger
+    let executor: ShellExecuting
+
+    init(logger: Logger,
+         fileReader: FileReading = FileReader(),
+         executor: ShellExecuting = ShellExecutor()) {
+        self.fileReader = fileReader
+        self.logger = logger
+        self.executor = executor
+    }
+
+    enum Errors: Error {
+        case failedToResolveName(URL)
+        case failedToReadPackageFile(String)
+        case failedToResolveLatestVersion(URL)
+    }
+
+    func nameOfPackage(at url: URL, temporaryFolder: String) throws -> String {
+        do {
+            guard !url.isForRemoteRepository else {
+                return try nameOfRemotePackage(at: url, temporaryFolder: temporaryFolder)
+            }
+
+            let folder = url.absoluteString
+            return try nameOfPackage(in: folder)
+        } catch {
+            throw Errors.failedToResolveName(url)
+        }
+    }
+
+    func latestMajorVersionForPackage(at url: URL) throws -> Int {
+        guard let releases = try? versions(for: url),
+            let latestVersion = releases.sorted().last else {
+            throw Errors.failedToResolveLatestVersion(url)
+        }
+
+        logger.logInfo("Using \(url.absoluteString) latest major:\(latestVersion.major)")
+
+        return latestVersion.major
+    }
+
+    func resolvePinnedPackages(generatedFolder: String) throws -> [Package.Pinned] {
+        struct ResolvedPackagesState: Decodable {
+            struct Object: Decodable {
+                let pins: [Package.Pinned]
+            }
+
+            let object: Object
+        }
+
+        let data = try fileReader.readData(atPath: generatedFolder.appendingPath("Package.resolved"))
+        let state: ResolvedPackagesState = try data.decoded()
+
+        return state.object.pins
+    }
+
+    private func nameOfRemotePackage(at url: URL, temporaryFolder: String) throws -> String {
+        removeCloneFolder(temporaryFolder: temporaryFolder)
+
+        logger.logInfo("Cloning \(url.absoluteString)...", isVerbose: true)
+
+        let clone = temporaryFolder.appendingPath("Clone")
+        try executor.spawn("git clone", arguments: ["\(url.absoluteString)", "--single-branch", "--depth 1", "\(clone)", "-q"])
```

Comment:
<!--
  0 failure: 
  1 warning:  Line should be 12...
  
  
  DangerID: danger-id-Danger;
  File: Sources/DangerDependenciesResolver/PackageDataProvider.swift;
  Line: 76;
-->

- :warning: Line should be 120 characters or less: currently 127 characters (`line_length`)